// xmind-model/dist/common/constants/models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MODEL_TYPE = {
    BASE_COMPONENT: 'BaseComponent',
    STYLE_COMPONENT: 'StyleComponent',
    STYLE: 'Style',
    BOUNDARY: 'Boundary',
    SUMMARY: 'Summary',
    HREF: 'Href',
    LABEL: 'Label',
    LEGEND: 'Legend',
    MARKER: 'Marker',
    NOTE: 'Note',
    NUMBERING: 'Numbering',
    RELATIONSHIP: 'Relationship',
    SHEET: 'Sheet',
    THEME: 'Theme',
    TOPIC: 'Topic',
    EXTENSION: 'Extension',
    IMAGE: 'Image',
    WORKBOOK: 'workbook'
};
exports.TOPIC_TYPE = {
    ATTACHED: 'attached',
    DETACHED: 'detached',
    SUMMARY: 'summary',
    CALLOUT: 'callout',
    ROOT: 'root'
};
exports.NUMBERFORMAT = {
    "NONE": "org.xmind.numbering.none",
    "ARABIC": "org.xmind.numbering.arabic",
    "ROMAN": "org.xmind.numbering.roman",
    "LOWERCASE": "org.xmind.numbering.lowercase",
    "UPPERCASE": "org.xmind.numbering.uppercase"
};
exports.NUMBERSEPARATOR = {
    "COMMA": "org.xmind.numbering.separator.comma",
    "DOT": "org.xmind.numbering.separator.dot",
    "HYPHEN": "org.xmind.numbering.separator.hyphen",
    "DASH": "org.xmind.numbering.separator.dash",
    "OBLIQUE": "org.xmind.numbering.separator.oblique"
};
exports.INFO_ITEM_DISPLAY_MODE = {
    CARD: 'card',
    ICON: 'icon'
};
exports.INFOITEM_TYPE_FULL = {
    "LABEL": "org.xmind.ui.infoItem.label",
    "HREF": "org.xmind.ui.infoItem.hyperlink",
    "NOTE": "org.xmind.ui.infoItem.notes",
    "TASK": "org.xmind.ui.infoItem.taskInfo",
    "AUDIO": "org.xmind.ui.infoItem.AudioNotes",
    "COMMENT": "org.xmind.ui.infoItem.comments"
};
exports.INFOITEM_TYPE_SHORT = {
    "LABEL": "label",
    "HREF": "href",
    "NOTE": "note",
    "TASK": "task",
    "AUDIO": "audio",
    "COMMENT": "comments"
};
exports.INFO_ITEM_STYLE_TYPE = {
    CLASSIC: 'classic',
    FASHION: 'fashion',
    // 根据json格式
    ACC_TO_JSON: 'accToJSON'
};
exports.COMPONENT_TYPE = exports.MODEL_TYPE;
exports.ATTACHMENT_PREFIX = "xap:resources/";
exports.TITLE_MAX_WIDTH = 300;
exports.TOPIC_TITLE_MAX_WIDTH = 300;
exports.TOPIC_MAX_CUSTOM_WIDTH = 762;
exports.TOPIC_DEFAULT_STRUCTURE = "org.xmind.ui.logic.right";
exports.DIRECTION = {
    UP: "up",
    DOWN: "down",
    LEFT: "left",
    RIGHT: "right",
    NONE: "none"
};
/**
 * @deprecated
 */
exports.TOPIC_ATTACHED = 'attached';
/**
 * @deprecated
 */
exports.TOPIC_DETACHED = 'detached';
/**
 * @deprecated
 */
exports.TOPIC_SUMMARY = 'summary';
/**
 * @deprecated
 */
exports.TOPIC_CALLOUT = 'callout';
/**
 * @deprecated
 */
exports.TOPIC_ROOT = 'root';
/**
* @description master range name
* */
exports.MASTER_RANGE = 'master';
exports.MANIFEST = {
    FILE_ENTRIES: 'file-entries',
    FILE_ENTRY: 'file-entry'
};
