// xmind-model/dist/models/base-component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// @flow
const constants_1 = require("../common/constants");
const base_model_1 = require("./base-model");
class BaseComponent extends base_model_1.default {
    constructor() {
        super(...arguments);
        this.componentType = constants_1.COMPONENT_TYPE.BASE_COMPONENT;
    }
    init(sheet) {
        this.ownerSheet(sheet);
    }
    triggerModelChanged(...args) {
        this.emit(args[0], ...args.slice(1));
        this.ownerSheet().emit(constants_1.EVENTS.AFTER_SHEET_CONTENT_CHANGE);
    }
    ownerWorkbook(workbook) {
        if (workbook)
            this._ownerWorkbook = workbook;
        return this._ownerWorkbook;
    }
    ownerSheet(sheet) {
        if (sheet)
            this._ownerSheet = sheet;
        return this._ownerSheet;
    }
    parent(parentModel) {
        if (parentModel !== undefined) {
            this.emit('beforeParentChange');
            this._parent = parentModel;
            this.emit('afterParentChange');
        }
        return this._parent;
    }
    getUndo() {
        return this.ownerSheet().getUndo();
    }
    isOrphan() {
        const parent = this.parent();
        if (parent) {
            return parent.isOrphan();
        }
        return true;
    }
}
exports.default = BaseComponent;
