// xmind-model/dist/models/range-component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("../common/constants");
const style_component_1 = require("./style-component");
const DATA_KEY_RANGE = 'range';
const GET = 'get';
const SET = 'set';
class RangeComponent extends style_component_1.default {
    getRange() {
        return this._data.commit(GET, DATA_KEY_RANGE);
    }
    setRange(range) {
        const oldRange = this.getRange();
        if (range === oldRange) {
            return;
        }
        this._data.commit(SET, DATA_KEY_RANGE, range);
        this.getUndo().add({
            undo: () => this.setRange(oldRange),
            redo: () => this.setRange(range)
        });
    }
    get rangeStart() {
        const range = this.getRange();
        if (range === constants_1.MASTER_RANGE)
            return -1;
        return parseInt(range.match(/\d+/g)[0], 10);
    }
    get rangeEnd() {
        const range = this.getRange();
        if (range === constants_1.MASTER_RANGE)
            return -1;
        return parseInt(range.match(/\d+/g)[1], 10);
    }
}
exports.default = RangeComponent;
