// xmind-model/dist/models/topic-image.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("../common/constants");
const base_component_1 = require("./base-component");
class TopicImage extends base_component_1.default {
    constructor() {
        super(...arguments);
        this.componentType = constants_1.COMPONENT_TYPE.IMAGE;
    }
    getSrc() {
        return this.get('src');
    }
    getWidth() {
        return this.get('width');
    }
    getHeight() {
        return this.get('height');
    }
    getAlign() {
        return this.get('align');
    }
    resize(newSize) {
        const oldWidth = this.getWidth();
        const oldHeight = this.getHeight();
        this._resize(newSize);
        this.getUndo().add({
            undo: () => this._resize({ width: oldWidth, height: oldHeight }),
            redo: () => this._resize(newSize)
        });
        return this;
    }
    _resize(newSize) {
        this.set('width', newSize.width);
        this.set('height', newSize.height);
    }
    align(newAlign) {
        const oldAlign = this.getAlign();
        this._align(newAlign);
        this.getUndo().add({
            undo: () => this._align(oldAlign),
            redo: () => this._align(newAlign)
        });
        return this;
    }
    _align(newAlign) {
        this.set('align', newAlign);
    }
    removeSelf() {
        const topic = this.parent();
        this.parent(null);
        topic.removeImage();
    }
}
exports.default = TopicImage;
